File: algorithms/lisp_interpreter/lisp.rb
#!/usr/bin/env ruby
# A minimal Lisp interpreter in Ruby
# Inspired by Peter Norvig's legendary implementation

# Tokenize
def tokenize(chars)
  chars.gsub('(', ' ( ').gsub(')', ' ) ').split
end

# Parse
def parse(program)
  read_from_tokens(tokenize(program))
end

def read_from_tokens(tokens)
  raise SyntaxError, 'unexpected EOF' if tokens.empty?
  
  token = tokens.shift
  
  if token == '('
    list = []
    list << read_from_tokens(tokens) while tokens[0] != ')'
    tokens.shift # pop off ')'
    list
  elsif token == ')'
    raise SyntaxError, 'unexpected )'
  else
    atom(token)
  end
end

def atom(token)
  # Try integer
  return token.to_i if token =~ /^-?\d+$/
  # Try float
  return token.to_f if token =~ /^-?\d+\.\d+$/
  # Symbol
  token.to_sym
end

# Environment
class Env
  attr_reader :data, :outer
  
  def initialize(params = [], args = [], outer = nil)
    @data = Hash[params.zip(args)]
    @outer = outer
  end
  
  def find(var)
    @data.key?(var) ? self : @outer&.find(var)
  end
  
  def [](var)
    find(var)&.data&.[](var)
  end
  
  def []=(var, val)
    @data[var] = val
  end
end

def standard_env
  env = Env.new
  env.data.merge!({
    :+ => ->(* args) { args.reduce(0, :+) },
    :- => ->(*args) { args.size == 1 ? -args[0] : args.reduce(:-) },
    :* => ->(*args) { args.reduce(1, :*) },
    :/ => ->(*args) { args.reduce(:/) },
    :< => ->(a, b) { a < b },
    :> => ->(a, b) { a > b },
    :<= => ->(a, b) { a <= b },
    :>= => ->(a, b) { a >= b },
    :'=' => ->(a, b) { a == b },
    :abs => ->(x) { x.abs },
    :append => ->(*args) { args.flatten(1) },
    :apply => ->(proc, args) { proc.call(*args) },
    :begin => ->(*args) { args.last },
    :car => ->(x) { x.first },
    :cdr => ->(x) { x.drop(1) },
    :cons => ->(x, y) { [x] + y },
    :'eq?' => ->(a, b) { a.equal?(b) },
    :'equal?' => ->(a, b) { a == b },
    :length => ->(x) { x.length },
    :list => ->(*args) { args },
    :'list?' => ->(x) { x.is_a?(Array) },
    :map => ->(fn, arr) { arr.map { |x| fn.call(x) } },
    :max => ->(*args) { args.max },
    :min => ->(*args) { args.min },
    :not => ->(x) { !x },
    :'null?' => ->(x) { x.is_a?(Array) && x.empty? },
    :'number?' => ->(x) { x.is_a?(Numeric) },
    :print => ->(*args) { puts args.join(' ') },
    :'procedure?' => ->(x) { x.respond_to?(:call) },
    :round => ->(x) { x.round },
    :'symbol?' => ->(x) { x.is_a?(Symbol) },
  })
  env
end

$global_env = standard_env

# Eval
def lisp_eval(x, env = $global_env)
  if x.is_a?(Symbol)                    # variable reference
    env[x] || (raise "undefined symbol: #{x}")
  elsif !x.is_a?(Array)                 # constant
    x
  else
    op, *args = x
    
    case op
    when :quote                         # quotation
      args[0]
    when :if                            # conditional
      test, conseq, alt = args
      exp = lisp_eval(test, env) ? conseq : alt
      lisp_eval(exp, env)
    when :define                        # definition
      var, exp = args
      env[var] = lisp_eval(exp, env)
    when :lambda                        # procedure
      params, body = args
      ->(*vals) { lisp_eval(body, Env.new(params, vals, env)) }
    else                                # procedure call
      proc = lisp_eval(op, env)
      vals = args.map { |arg| lisp_eval(arg, env) }
      proc.call(*vals)
    end
  end
end

def schemestr(exp)
  if exp.is_a?(Array)
    '(' + exp.map { |e| schemestr(e) }.join(' ') + ')'
  else
    exp.to_s
  end
end

# REPL
def repl(prompt = 'lisp> ')
  loop do
    print prompt
    input = gets
    break unless input
    
    input = input.strip
    next if input.empty?
    
    begin
      val = lisp_eval(parse(input))
      puts schemestr(val) unless val.nil?
    rescue Interrupt, SystemExit
      puts "\nBye!"
      break
    rescue => e
      puts "Error: #{e.message}"
    end
  end
  
  puts "\nBye!" if $stdin.tty?
end

# Run REPL if this is the main script
repl if __FILE__ == $PROGRAM_NAME

File: algorithms/lisp_interpreter/lisp.js
#!/usr/bin/env node
/**
 * A minimal Lisp interpreter in JavaScript
 * Inspired by Peter Norvig's legendary implementation
 */

const readline = require('readline');

// Tokenize
function tokenize(chars) {
    return chars.replace(/\(/g, ' ( ').replace(/\)/g, ' ) ').trim().split(/\s+/);
}

// Parse
function parse(program) {
    return readFromTokens(tokenize(program));
}

function readFromTokens(tokens) {
    if (tokens.length === 0) throw new SyntaxError('unexpected EOF');
    const token = tokens.shift();
    if (token === '(') {
        const L = [];
        while (tokens[0] !== ')') {
            L.push(readFromTokens(tokens));
        }
        tokens.shift(); // pop off ')'
        return L;
    } else if (token === ')') {
        throw new SyntaxError('unexpected )');
    } else {
        return atom(token);
    }
}

function atom(token) {
    const num = Number(token);
    return isNaN(num) ? token : num;
}

// Environment
class Env {
    constructor(parms = [], args = [], outer = null) {
        this.dict = {};
        for (let i = 0; i < parms.length; i++) {
            this.dict[parms[i]] = args[i];
        }
        this.outer = outer;
    }
    
    find(varName) {
        return (varName in this.dict) ? this : this.outer.find(varName);
    }
    
    get(varName) {
        return this.find(varName).dict[varName];
    }
    
    set(varName, value) {
        this.dict[varName] = value;
    }
}

function standardEnv() {
    const env = new Env();
    env.dict = {
        '+': (...args) => args.reduce((a, b) => a + b, 0),
        '-': (...args) => args.length === 1 ? -args[0] : args.reduce((a, b) => a - b),
        '*': (...args) => args.reduce((a, b) => a * b, 1),
        '/': (...args) => args.reduce((a, b) => a / b),
        '>': (a, b) => a > b,
        '<': (a, b) => a < b,
        '>=': (a, b) => a >= b,
        '<=': (a, b) => a <= b,
        '=': (a, b) => a === b,
        'abs': Math.abs,
        'append': (...args) => args.flat(),
        'apply': (proc, args) => proc(...args),
        'begin': (...args) => args[args.length - 1],
        'car': (x) => x[0],
        'cdr': (x) => x.slice(1),
        'cons': (x, y) => [x, ...y],
        'eq?': (a, b) => a === b,
        'equal?': (a, b) => JSON.stringify(a) === JSON.stringify(b),
        'length': (x) => x.length,
        'list': (...args) => args,
        'list?': (x) => Array.isArray(x),
        'map': (fn, arr) => arr.map(fn),
        'max': Math.max,
        'min': Math.min,
        'not': (x) => !x,
        'null?': (x) => Array.isArray(x) && x.length === 0,
        'number?': (x) => typeof x === 'number',
        'print': (...args) => console.log(...args),
        'procedure?': (x) => typeof x === 'function',
        'round': Math.round,
        'symbol?': (x) => typeof x === 'string',
    };
    return env;
}

const globalEnv = standardEnv();

// Eval
function evaluate(x, env = globalEnv) {
    if (typeof x === 'string') {         // variable reference
        return env.get(x);
    } else if (!Array.isArray(x)) {      // constant
        return x;
    }
    const [op, ...args] = x;
    if (op === 'quote') {                // quotation
        return args[0];
    } else if (op === 'if') {            // conditional
        const [test, conseq, alt] = args;
        const exp = evaluate(test, env) ? conseq : alt;
        return evaluate(exp, env);
    } else if (op === 'define') {        // definition
        const [symbol, exp] = args;
        env.set(symbol, evaluate(exp, env));
    } else if (op === 'lambda') {        // procedure
        const [parms, body] = args;
        return (...args) => evaluate(body, new Env(parms, args, env));
    } else {                             // procedure call
        const proc = evaluate(op, env);
        const vals = args.map(arg => evaluate(arg, env));
        return proc(...vals);
    }
}

function schemestr(exp) {
    if (Array.isArray(exp)) {
        return '(' + exp.map(schemestr).join(' ') + ')';
    }
    return String(exp);
}

// REPL
function repl() {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
        prompt: 'lisp> '
    });
    
    rl.prompt();
    
    rl.on('line', (line) => {
        try {
            const val = evaluate(parse(line));
            if (val !== undefined) {
                console.log(schemestr(val));
            }
        } catch (e) {
            console.log('Error:', e.message);
        }
        rl.prompt();
    });
    
    rl.on('close', () => {
        console.log('\nBye!');
        process.exit(0);
    });
}

if (require.main === module) {
    repl();
}

module.exports = { tokenize, parse, evaluate, Env, standardEnv };

